tests/phpunit/tests/admin/includesMedia.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+/**
+ * @group media
+ * @group admin
+ */
+class Tests_Admin_includesMedia extends WP_UnitTestCase {
+	/** @var \WP_Post */
+	protected static $post;
+	/** @var \WP_Post */
+	protected static $different_post;
+
+	public static function wpSetUpBeforeClass( \WP_UnitTest_Factory $factory ) {
+		self::$post           = $factory->post->create_and_get();
+		self::$different_post = $factory->post->create_and_get();
+	}
+
+	public function wpTearDownAfterClass() {
+		self::$post           = null;
+		self::$different_post = null;
+	}
+
+	public function setUp() {
+		parent::setUp();
+
+		add_filter( 'pre_http_request', [ $this, 'internal_fake_download_url' ], 10, 3 );
+	}
+
+	public function tearDown() {
+		parent::tearDown();
+
+		$this->remove_added_uploads();
+		remove_filter( 'pre_http_request', [ $this, 'internal_fake_download_url' ] );
+	}
+
+	/**
+	 * Filters whether to preempt an HTTP request's return value. Mocking the
+	 * response. Since ticket 49631
+	 *
+	 * @param false  $preempt Whether to preempt an HTTP request's return value. Default false.
+	 * @param array  $parsed_args HTTP request arguments.
+	 * @param string $url The request URL.
+	 *
+	 * @return array
+	 */
+	public function internal_fake_download_url( $preempt, array $parsed_args, $url = '' ) {
+		// Just need an image content string to fill the new image, keeping the same mime type.
+		file_put_contents( $parsed_args['filename'], file_get_contents( DIR_TESTDATA . '/images/canola.jpg' ) );
+
+		return [
```

Comment:
This seems incorrect to me. Based on my reading on where this filter should be called, I think that this would need to return more than just a response code in order to test this accurately unless I'm missing something.

Reply:
The hook could return a:

```php
// An array containing 'headers', 'body', 'response', 'cookies', and 'filename' elements
```

Then the function `download_url` does need a 200 response only to continue.

```php
	if ( 200 != $response_code ) {
```
So only by shocking (applying a short-circuit as class-http.php said) on the request I mock a 200 response and just copy the content into the tmp-name file where the process will continue and save the attachment and added to the post.